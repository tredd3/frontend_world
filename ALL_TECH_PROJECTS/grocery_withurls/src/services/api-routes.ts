const baseUrl = process.env.REACT_APP_BASE_URL || '';
const baseUrlLogin = process.env.REACT_APP_BASE_URL_LOGIN || '';

export const apiRoutes = {
  getNewAccessToken: `${baseUrl}/jm/auth/oauth/v2/token`,
  getDashboard: `${baseUrl}/jiokart/v1/jiogrocery/get/dashboard`,
  viewAllDashboard: `${baseUrl}/jiokart/v1/jiogrocery/view/all/dashboard`,
  getStores: `${baseUrl}/jiokart/v1/get/stores`,
  getCart: `${baseUrl}/jiokart/v1/jiogrocery/get/cart`,
  addCartItem: `${baseUrl}/jiokart/v1/add/cart/item`,
  removeCartItem: `${baseUrl}/jiokart/v1/remove/cart/item`,
  moveCartItem: `${baseUrl}/jiokart/v1/jiogrocery/move/item`,
  getCategories: `${baseUrl}/jiokart/v1/get/categories`,
  getProductList: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  checkServiceability: `${baseUrl}/jiokart/v1/jiogrocery/pincode/check`,
  getProductSearch: `${baseUrl}/jiokart/v1/jiogrocery/product/search`,
  getSuggestions: `${baseUrl}/jiokart/v1/jiogrocery/suggestion`,
  addUser: `${baseUrl}/jiokart/v1/jiogrocery/add/user`,
  saveAddress: `${baseUrl}/jiokart/v1/jiogrocery/save/user/address`,
  getAddress: `${baseUrl}/jiokart/v1/jiogrocery/get/user/address`,
  deleteAddress: `${baseUrl}/jiokart/v1/jiogrocery/remove/user/address`,
  getProductDetails: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  getSimilarItems: `${baseUrl}/jiokart/v1/jiogrocery/store/similar/items`,
  getOrderDetails: `${baseUrl}/jiokart/v1/jiogrocery/order/details`,
  getOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/listing`,
  getDeliveryModes: `${baseUrl}/jiokart/v1/get/delivery/modes`,
  getLocation: `${baseUrl}/jiokart/v1/jiogrocery/google/places`,
  searchOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/item/search`,
  checkout: `${baseUrl}/jiokart/v1/order/checkout`,
  placeOrder: `${baseUrl}/jiokart/v1/place/order`,
  getUserDetail: `${baseUrl}/jiokart/v1/jiogrocery/get/user`,
  getWishListData: `${baseUrl}/jiokart/v1/jiogrocery/get/item/list`,
  addItemToWishList: `${baseUrl}/jiokart/v1/jiogrocery/add/item`,
  removeItemFromWishList: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`,
  sendOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/send`,
  verifyOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/verify`,
  validateuser: `${baseUrl}/cr/v1/validateuser`,
  cancelOrder: `${baseUrl}/jiokart/v1/order/cancel`,
  config: `${baseUrl}/jiokart/v1/logged/customer/configurations`,
  updateUser: `${baseUrl}/jiokart/v1/jiogrocery/update/user`,
  orderRating: `${baseUrl}/jiokart/v1/order/rating`,
  getPaymentRequest: `${baseUrl}/jiokart/v1/get/payment/request`,
  getPaymentResponse: `${baseUrl}/jiokart/v1/get/payment/response`,
  getNotifications: `${baseUrl}/jiokart/v1/jiogrocery/notification/get`,
  saveNotification: `${baseUrl}/jiokart/v1/jiogrocery/notification/save`,
  clearNotifications: `${baseUrl}/jiokart/v1/jiogrocery/clear/notification`,
  removeCartSavedItem: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`
} as const;

const endpointsWithoutAuth = [
  apiRoutes.checkServiceability,
  apiRoutes.getStores,
  apiRoutes.getProductDetails,
  apiRoutes.getSimilarItems,
  apiRoutes.getLocation,
  apiRoutes.sendOtp,
  apiRoutes.getDashboard
];

const endpointsNeedingEnvelope = [
  apiRoutes.getCategories,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.getStores,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.getPaymentRequest,
  apiRoutes.getPaymentResponse,
  apiRoutes.orderRating
];

const endpointsNeedingStoreId = [
  apiRoutes.getCart,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.placeOrder,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.saveAddress,
  apiRoutes.getAddress,
  apiRoutes.getPaymentResponse,
  apiRoutes.removeCartSavedItem
];

export const needsEnvelope = (url: string) => endpointsNeedingEnvelope.includes(url);
export const needsAuthentication = (url: string) => !endpointsWithoutAuth.includes(url);
export const needsStoreId = (url: string) => endpointsNeedingStoreId.includes(url);

export const v2ApiRoutes = [
  apiRoutes.getPaymentResponse,
  apiRoutes.getOrderHistory,
  apiRoutes.getOrderDetails,
  apiRoutes.getCart,
  apiRoutes.getDeliveryModes
];
