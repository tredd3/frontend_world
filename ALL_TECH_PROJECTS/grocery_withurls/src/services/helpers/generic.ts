import { APIUser, UserPivot } from '../../types';

export const genderCodeToString = (code: APIUser['gender']): UserPivot['gender'] => {
  if (code === undefined) return undefined;
  return code === 1 ? 'male' : 'female';
};

export const genderStringToCode = (genderString: UserPivot['gender']): APIUser['gender'] => {
  if (genderString === undefined || genderString.length === 0) return undefined;
  return genderString === 'male' ? 1 : 2;
};
