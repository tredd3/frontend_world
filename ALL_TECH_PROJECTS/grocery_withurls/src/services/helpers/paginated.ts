const mergeArrays = <T>(previousArray: T[], nextArray: T[]): T[] => (
  [...previousArray, ...nextArray]
);

export type Paginated<T> = {
  data: T | null;
  loadNextPage: null | (() => Promise<Paginated<T>>);
}

// So admittedly, this function looks complicated. But that's only because
// there's a ton of annotations to make it typesafe. The usage is actually
// rather simple. See the tests to see detailed examples.
// This function has two overloads - one that returns arrays, and one that
// returns any other type.
//
// When returning arrays:
//   ```
//   const paginatedService = paginated(
//     (pageNumber = 0, arg1: num, arg2: str) => {
//       return {
//         data: await makeApiCall(/* ... */), // returns Promise<Array<T>>
//         nextPage: pageNumber + 1
//       }
//     }
//   );
//   ```
// Now, `paginatedService`, which will be used in the app, has the
// following signature:
// `(arg1: num, arg2: str) => Promise<{ data: Array<T>, nextPage: null | fn }>`
// That is, you invoke it with the args, and await for data and nextPage.
// `nextPage` is null or a fn. If null, we're done with pagination.
// If fn, we can get more data by calling the fn, which continues to return
// the same type Promise<{ data: Array<T>, nextPage: null | fn }>.
// `data` contains the cumulative data from all pages, so you don't have to
// aggregate it yourself in component state.
//
// When returning non-arrays:
//
//   ```
//   const paginatedService = paginated(
//     (pageNumber = 0, arg1: num, arg2: str) => {
//       return {
//         data: await makeApiCall(/* ... */), // return Promise<any !extends []>
//         nextPage: pageNumber + 1
//       }
//     },
//     (a, b) => { /* merge a and b */}
//   );
//   ```
// The service call itself is similar to the previous example, except
// makeApiCall doesn't return an array in this case. This is the case with
// say getKiranas. In such a situation, you also need to specify how to merge
// previous results with the next results, which you do by providing a second
// `merge` function. The return type of invoking `paginatedService` is the
// same as above, except data will obviously not be an array.
//
// When you don't have any more data to return, simply return null.
//
// In both examples above, the `pageNumber` above is specified by as a number
// but that's not necessary. In fact, it could be of any type, as long as the
// `nextPage` has the same type as `pageNumber`. This is allow you to have
// different pagination parameters per page. (eg. last offset, offset size).

function paginated<P, A extends any[], R>(
  serviceCall: (pageNum: P, ...args: A) =>
    Promise<null | { nextPage: P; data: R[] }>
): (...args: A) => Promise<Paginated<R[]>>
function paginated<P, A extends any[], R>(
  serviceCall: (pageNum: P, ...args: A) =>
    Promise<null | { nextPage: P; data: R }>,
  merge: (a: R, b: R) => R
): (...args: A) => Promise<Paginated<R>>
function paginated <P, A extends any[], R>(
  serviceCall: (pageNum: P, ...args: A) =>
    Promise<null | { nextPage: P; data: R[] | R }>,
  merge?: (a: R, b: R) => R
): (...args: A) => Promise<Paginated<R | R[]>> {
  let pagination: P;
  let data: R | R[] | null = null;

  return (...args: A): Promise<Paginated<R | R[]>> => {
    // Reset internal state incase we're loading this for the first time.
    // This is needed so that if a user goes forward from a page with a listing
    // and comes back, we need to ensure that our internal state is reset.
    if (pagination === null) data = null;

    const loadNextPage = async (): Promise<Paginated<R | R[]>> => {
      const result = await serviceCall(pagination, ...args);

      if (result === null) return { data, loadNextPage: null };
      if (!Array.isArray(result.data) && !merge) {
        throw new Error('You must provide a merge function for non-array return types');
      }

      pagination = result.nextPage;

      if (data === null) { // We don't have any data so far
        data = result.data;
      } else if (Array.isArray(result.data)) { // First overload: array return
        data = mergeArrays((data || []) as R[], result.data);
      } else { // Second overload: non-array return
        data = data ? merge!(data as R, result.data) : result.data;
      }

      return { data, loadNextPage: pagination ? loadNextPage : null };
    };

    return loadNextPage();
  };
}

export default paginated;
