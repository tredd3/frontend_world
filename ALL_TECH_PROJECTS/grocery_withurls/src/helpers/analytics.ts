import { getMobileOperatingSystem } from './utilities';
import { getUser } from '../services/user';
import { getCart, PaymentMethodId, paymentLabelById } from '../services/cart';

import { CheckoutData } from '../services/checkout';
import { CartPivot } from '../types';
import { flatMap } from './typed-utils';

let purchaseJourney: string;

export const setPurchaseJourney = (val: string) => {
  purchaseJourney = val;
};

export const getPurchaseJourney = () => purchaseJourney;

declare global {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    digitalData: Record<string, any>;
    _satellite: {
      track: (event: string) => void;
    };
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const track = async (event: string, data: Record<string, any>) => {
  if (typeof window === 'undefined') return;
  const { userId } = await getUser();
  const userOS = getMobileOperatingSystem();
  const platform = userOS === 2 ? 'Android App' : userOS === 3 ? 'iOS App' : 'Web';

  const digitalData = {
    link: {
      ...data.link
    },
    internal: {
      ...data.internal
    },
    page: {
      pageInfo: {
        appName: 'JioMart Web'
      },
      ...data.page
    },
    user: {
      platform,
      loginStatus: 'Logged in',
      martId: userId,
      jioCenter: '',
      ...data.user
    },
    product: [
      ...(data.product ? data.product : [])
    ],
    interPromotion: {
      ...data.interPromotion
    },
    ...(data.product ? data.filterSection : {}),
    ...(data.purchase ? data.purchase : {})
  };

  window.digitalData = {
    ...window.digitalData,
    ...digitalData
  };

  // eslint-disable-next-line no-underscore-dangle
  if (window._satellite) window._satellite.track(event);
};

type LinkPositions = 'Top' | 'Middle' | 'Bottom';

export const trackLink = (linkName: string, linkType: string, linkPosition: LinkPositions) => (
  track('linkTracking', { link: { linkName, linkType, linkPosition } })
);

export const trackPage = (pageName: string) => (
  track('pageLoad', { page: { pageName } })
);

type FilterSection = {
  filterSelection: boolean;
  sortSelection: boolean;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const trackProduct = (eventName: string, productData: Record<string, any>, filterSection?: FilterSection) => {
  const defaultFilterSection = {
    filterSelection: false,
    sortSelection: false
  };
  return track(eventName, {
    product: [productData],
    filterSection: (filterSection || defaultFilterSection)
  });
};

export const trackAPIShipments = (eventName: string, shipments: CheckoutData['Shipments'], purchase?: any) => {
  if (shipments && shipments.length) {
    track(eventName, {
      product: flatMap(shipments, shipment => shipment.Product)
        .map(product => ({
          productID: product.ProductSkuId,
          quantity: product.SelectedQuantity,
          price: product.SelectedQuantity * product.SP,
          productName: product.ProductName
        })),
      ...(purchase && { purchase })
    });
  }
};

type PurchaseArg = {
  paymentMethod: PaymentMethodId;
  paymentStatus?: 'Paid' | 'Unpaid';
  cart: CartPivot;
  failureReason?: string;
}

export const trackShipments = async (eventName: string, purchase?: PurchaseArg) => {
  const { shipments } = purchase?.cart || await getCart();
  if (!shipments || !shipments.length) return Promise.resolve();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let additionalInfo: Record<string, any> | null = null;

  if (purchase) {
    const { paymentMethod, cart, ...rest } = purchase;
    additionalInfo = {
      paymentMethod: paymentLabelById(paymentMethod),
      couponDiscount: cart.couponDiscountAmount,
      deliveryCharges: cart.deliveryCharge,
      totalOrderValue: cart.totalPayableAmount,
      transactionId: cart.cartId,
      ...rest
    };
  }

  return track(eventName, {
    product: flatMap(shipments, shipment => shipment.products)
      .map(product => ({
        productId: product.skuId,
        quantity: product.selectedQuantity,
        price: product.selectedQuantity * product.sp,
        productName: product.name
      })),
    ...(additionalInfo && additionalInfo)
  });
};
