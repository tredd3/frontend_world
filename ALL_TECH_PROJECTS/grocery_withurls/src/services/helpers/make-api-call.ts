/* eslint-disable no-console */
import uuidV1 from 'uuid/v1';
import { needsEnvelope, needsAuthentication, v2ApiRoutes } from '../api-routes';
import { getTokens } from '../../intents';
import { HTTPError, APIError } from '../types';

export type ControlType = { Status: number; MessageCode: number; Message: string };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const prepareBody = <RequestBodyType extends Record<string, any>>(body: RequestBodyType, url: string) => {
  const requestBody = {
    Request: {
      Header: {
        DeviceId: '6cf35d17989d3fbe',
        RequestTime: new Date().getTime(),
        RequestId: uuidV1(),
        Version: v2ApiRoutes.includes(url) ? 2.0 : parseFloat(process.env.REACT_APP_API_VERSION!),
        Product: Number(process.env.REACT_APP_PRODUCT_VERSION)
      },
      Body: body || {}
    }
  };

  return needsEnvelope(url) ? { Data: requestBody } : requestBody;
};

type LogType = 'success' | 'error';
type LogInfo = {
  startTime: number;
  url: string;
  fetchOptions: any;
  response: any;
  json?: any;
  type: LogType;
};
const log = ({
  startTime, url, fetchOptions, response, json, type
}: LogInfo) => {
  if (process.env.NODE_ENV === 'development') { // Compiles this block out in non-dev builds
    const { body, ...otherOptions } = fetchOptions;
    console.groupCollapsed(
      `HTTP request: ${url} %c${Date.now() - startTime}ms %c${type}`,
      'color: grey;',
      `color: ${type === 'success' ? 'green' : 'red'}`
    );
    console.log('Request body:', JSON.stringify(JSON.parse(body), null, 2));
    if (json) console.log('Response json', JSON.stringify(json, null, 2));
    console.log('Full request:', { body, ...otherOptions });
    console.log('Full response:', response);
    console.groupEnd();
  }
};

const doesBodyContainError = ({ Control }: { Control: ControlType }) => (
  !Control || Control.Status === 0 || Control.MessageCode === 404
);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MakeApiCallOptions<RequestBodyType extends Record<string, any>> =
  Partial<Omit<RequestInit, 'body'>> & { body: RequestBodyType };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default async <RequestBodyType extends Record<string, any>, ResponseType>(
  url: string,
  { headers, body, ...otherOptions }: MakeApiCallOptions<RequestBodyType> = { body: {} as RequestBodyType }
) => {
  const tokens = await getTokens();

  const startTime = Date.now();
  const fetchOptions = {
    method: 'POST',
    headers: {
      'X-API-Key': tokens.apiKey,
      'Content-Type': 'application/json',
      'X-REQUEST-ID': uuidV1(),
      'app-name': 'RJIL_JioKart',
      ...(needsAuthentication(url) && { Authorization: `Bearer ${tokens.accessToken}` }),
      ...headers
    },
    body: JSON.stringify(prepareBody<RequestBodyType>(body, url)),
    ...otherOptions
  };
  const response = await fetch(url, fetchOptions);

  if (!response.ok) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, type: 'error'
      });
    }
    throw new HTTPError(response);
  }
  const json = await response.json();

  if (doesBodyContainError(json)) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, json, type: 'error'
      });
    }
    throw new APIError(json);
  }

  if (process.env.NODE_ENV === 'development') {
    log({
      startTime, url, fetchOptions, response, json, type: 'success'
    });
  }

  return json.Data ? json.Data as ResponseType : json.Control as ResponseType;
};
