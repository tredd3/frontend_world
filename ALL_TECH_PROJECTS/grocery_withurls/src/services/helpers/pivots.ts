import {
  APIShipment, APIShipmentProduct, APICart, APIProduct, ShipmentProductPivot,
  CartPivot, ProductPivot, couponDiscount, APIDashboardProduct,
  DashboardProductPivot, APIUser, UserPivot, AddressPivot, ShipmentPivot,
  APIStorePickup, APIShipmentProductCoupon, DeliveryModePivot, ConfigPivot,
  APIConfig, LastOrderDetails, Notification, NotificationPivot
} from '../../types';
import { genderCodeToString } from './generic';
import { APIAddressRequest } from '../address';

export const pivotShipmentProductCoupon = (coupon: APIShipmentProductCoupon) => ({
  id: coupon.CouponId,
  code: coupon.CouponCode,
  discountAmount: coupon.Discount,
  title: coupon.Title,
  description: coupon.Description
});

export const pivotShipmentProduct = (product: APIShipmentProduct): ShipmentProductPivot => ({
  mrp: product.MRP,
  sp: product.SP,
  otherImages: product.ProductOtherImage,
  name: product.ProductName,
  image: product.ProductImage,
  skuId: product.ProductSkuId,
  id: product.ProductId,
  isFc: product.IsFc,
  categoryId: product.CategoryId,
  selectedQuantity: product.SelectedQuantity,
  savingsPercent: product.PricePercentage,
  coupon: product.Coupons && pivotShipmentProductCoupon(product.Coupons)
});

export const pivotShipment = (shipment: APIShipment): ShipmentPivot => ({
  type: shipment.Type,
  deliveryDate: shipment.DeliveryDate,
  totalAmount: shipment.TotalAmount,
  discountAmount: shipment.TotalDiscount,
  couponDiscountAmount: shipment.TotalCouponDiscount,
  totalPayable: shipment.TotalPay,
  deliveryCharge: shipment.DeliveryCharge,
  deliveryDay: shipment.DeliveryDay,
  creationDate: shipment.CreationDate,
  itemCount: shipment.TotalItem,
  products: shipment.Product.map(pivotShipmentProduct)
});

export const pivotProduct = (product: APIProduct): ProductPivot => ({
  mrp: product.MRP,
  sp: product.SP,
  name: product.ProductName,
  images: (product.ProductOtherImage && product.ProductOtherImage.length) ? product.ProductOtherImage : [product.ProductImage],
  id: product.ProductId,
  skuId: product.ProductSkuId,
  isFc: product.IsFc,
  category: {
    name: product.CategoryName,
    id: product.CategoryId
  },
  alias: product.ProductAlias,
  description:
    product.SkuBulletPoint && product.SkuBulletPoint.length
      ? `${product.Description}. ${product.SkuBulletPoint.reduce(
        (acc, cv) => `${acc}. ${cv}`
      )}`
      : product.Description,
  disclaimer: product.Disclaimer,
  bestBefore: product.BestBeforeDate ? new Date(Number(product.BestBeforeDate)) : null,
  isVeg: product.FoodType ? Boolean(product.FoodType.toLocaleLowerCase() === 'veg') : null,
  additionalInformation: product.AdditionalInfo && product.AdditionalInfo !== 'null' ? product.AdditionalInfo : '',
  manufacturerAddress: product.ManufacturerAddress && product.ManufacturerAddress !== 'null' ? product.ManufacturerAddress : '',
  coupon: product.Coupons && Object.keys(product.Coupons).length > 0 ? {
    title: product.Coupons.Title,
    description: product.Coupons.Description,
    discountAmount: product.Coupons.DiscountType === couponDiscount.percentageDiscount
      ? product.Coupons.DiscountMax
      : product.Coupons.DiscountAbsValue
  } : null,
  brandName: product.BrandName,
  deliverySlot: product.DeliverySlot,
  ingredients: product.Ingredient,
  nutritionalFacts: product.NutritionalFacts
});

export const pivotCartResponse = (cartData: APICart): CartPivot => ({
  cartId: cartData.CartId,
  totalAmount: cartData.TotalAmount,
  discountAmount: cartData.TotalMRPDiscount,
  deliveryCharge: cartData.TotalDeliveryCharge,
  itemCount: cartData.TotalItem,
  couponDiscountAmount: cartData.TotalCouponDiscount,
  totalPayableAmount: cartData.TotalPay,
  shipments: cartData.Shipments.map(pivotShipment),
  savedForLater: cartData.SaveLater?.length ? cartData.SaveLater.map(pivotShipmentProduct) : null
});

export const pivotDashboardProduct = (product: APIDashboardProduct): DashboardProductPivot => ({
  brandName: product.BrandName,
  id: product.ProductId,
  images: product.ProductOtherImage?.length ? product.ProductOtherImage : [product.ProductImage],
  isFc: product.IsFc,
  mrp: product.MRP,
  name: product.ProductName,
  skuId: product.ProductSkuId,
  sp: product.SP,
  category: {
    name: product.CategoryName,
    id: product.CategoryId
  },
  coupon: product.CouponResult && Object.keys(product.CouponResult).length > 0 ? {
    title: product.CouponResult.Title,
    description: product.CouponResult.Description,
    discountAmount: product.CouponResult.DiscountType === couponDiscount.percentageDiscount
      ? product.CouponResult.DiscountMax
      : product.CouponResult.DiscountAbsValue
  } : null
});

export const pivotAPIUser = (user: APIUser): UserPivot => {
  const {
    mobileNumber, emailId, gender, dob, userId, ...otherUserProps
  } = user;
  const { latitude: lat, longitude: lng, ...remainingPreferences } = otherUserProps.preferences || ({} as UserPivot['preferences']);

  return {
    phoneNumber: mobileNumber,
    email: emailId,
    userId,
    gender: genderCodeToString(gender),
    ...(dob && { dob: new Date(dob) }),
    ...otherUserProps,
    preferences: { ...remainingPreferences, lat, lng },
    login: (userId !== undefined && userId !== null)
  };
};

export const pivotAPIAddress = ({ addressId, addedDate, ...address }: APIAddressRequest): AddressPivot => ({
  id: addressId,
  addedDate: new Date(addedDate.replace(/(A|P)M IST/, '')),
  ...address
});

export const pivotDeliveryMode = (date: APIStorePickup): DeliveryModePivot => ({
  id: date.DateId,
  label: date.Date,
  value: date.DateValue,
  isOpen: date.OpenStatus === 1,
  store: {
    isKdp: Boolean(date.KdpStore),
    name: date.Store.Name,
    storeId: Number(date.Store.StoreId)
  }
});

export const pivotConfig = (apiConfig: APIConfig): ConfigPivot => {
  let lastOrderDetails = null;
  if (apiConfig.LAST_ORDER_DETAILS && Object.keys(apiConfig.LAST_ORDER_DETAILS).length !== 0) {
    const lod = apiConfig.LAST_ORDER_DETAILS as LastOrderDetails;
    lastOrderDetails = {
      totalItems: Number(lod.TOTAL_ITEMS),
      orderId: Number(lod.ORDER_ID),
      storeName: lod.STORE_NAME,
      totalShipment: lod.TOTAL_SHIPMENT,
      deliveredDate: lod.DELIVERED_DATE,
      storeId: lod.STORE_ID,
      orderAmount: lod.ORDER_AMOUNT,
      deliveredTime: lod.DELIVERED_TIME
    };
  }

  return {
    pageSize: Number(apiConfig.DEFAULT_PAGE_SIZE),
    quantityErrorMessage: apiConfig.QTY_ERROR_MESSAGE,
    invalidOTPMessage: apiConfig.INVALID_OTP,
    maxCODLimit: Number(apiConfig.MAX_COD_LIMIT),
    deliveryChargeThreshold: Number(apiConfig.DELIVERY_CHARGE_THRESHOLD_AMOUNT),
    unreadNotificationCount: Number(apiConfig.UNREAD_NOTIFICATION_COUNT),
    deliveryCharges: Number(apiConfig.DELIVERY_CHARGES),
    codLimitErrorMessage: apiConfig.COD_LIMIT_ERROR_MESSAGE,
    welcomeProductSKU: Number(apiConfig.WELCOME_PRODUCT_SKU),
    genericErrorMessage: apiConfig.GENERIC_ERROR_MESSAGE,
    maxQuantityLimitPerItem: Number(apiConfig.MAX_QTY_LIMIT_PER_ITEM),
    lastOrderDetails
  };
};

export const pivotNotification = (notification: Notification): NotificationPivot => ({
  id: notification.notificationId,
  title: notification.title,
  body: notification.body,
  orderId: notification.data?.orderId ? notification.data.orderId : undefined,
  date: new Date(notification.createdTime),
  icon: notification.icon,
  status: notification.status.toLowerCase() as 'read' | 'unread'
});
