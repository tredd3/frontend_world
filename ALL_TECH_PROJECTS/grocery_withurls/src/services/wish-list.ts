import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { pivotProduct } from './helpers/pivots';
import { APIProduct, ProductPivot } from '../types';

export const getWishList = async (): Promise<ProductPivot[]> => {
  const { Products } = await makeApiCall<{
    productCategoryType: 'Wishlist';
  }, {
    Products: APIProduct[];
  }>(apiRoutes.getWishListData, {
    body: {
      productCategoryType: 'Wishlist'
    }
  });

  return Products.map(pivotProduct);
};

export const removeFromWishList = async (skuId: number) => makeApiCall(apiRoutes.removeItemFromWishList, {
  body: {
    productCategoryId: '',
    productCategoryType: 'Wishlist',
    productType: 1,
    qty: 0,
    skuId
  }
});
