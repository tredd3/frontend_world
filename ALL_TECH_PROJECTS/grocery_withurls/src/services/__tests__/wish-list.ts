import { getWishList, removeFromWishList } from '../wish-list';
import * as makeApiCall from '../helpers/make-api-call';
import { apiRoutes } from '../api-routes';
import mockDB from '../../mocks/db.json';

describe('Wish list service tests', () => {
  let makeApiCallSpy: jest.SpyInstance<Promise<unknown>>;

  beforeAll(() => {
    makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.wishlist.Data);
  });

  afterEach(() => {
    makeApiCallSpy.mockClear();
  });

  it('should get items from wishlist correctly', async () => {
    const wishListData = await getWishList();
    expect(wishListData).toMatchSnapshot();
    expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
    expect(makeApiCallSpy).toHaveBeenLastCalledWith(apiRoutes.getWishListData, {
      body: {
        productCategoryType: 'Wishlist'
      }
    });
  });

  it('should remove item from wishlist correctly', async () => {
    const dummySkuId = 1234;
    await removeFromWishList(dummySkuId);

    expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
    expect(makeApiCallSpy).toHaveBeenLastCalledWith(apiRoutes.removeItemFromWishList, {
      body: {
        productCategoryId: '',
        productCategoryType: 'Wishlist',
        productType: 1,
        qty: 0,
        skuId: dummySkuId
      }
    });
  });
});
