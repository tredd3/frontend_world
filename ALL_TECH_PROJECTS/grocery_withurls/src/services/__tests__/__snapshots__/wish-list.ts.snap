// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wish list service tests should get items from wishlist correctly 1`] = `
Array [
  Object {
    "additionalInformation": "",
    "alias": "Kato Grape Juice with Nata De Coco 330 ml 1 pcs",
    "bestBefore": 2020-06-02T06:30:00.000Z,
    "brandName": "Kato",
    "category": Object {
      "id": 10001089,
      "name": "Fruit Juice",
    },
    "coupon": null,
    "deliverySlot": undefined,
    "description": "Feel refreshed with the heavenly taste of Kato Kato Grape Juice with Nata De Coco. It tastes amazing, gives you an instant energy kick and quenches your thirst. It is also made from the finest quality of grapes and combines quality and affordability at the same time. So whether you are a child or an adult, this grape juice will satisfy your Grape cravings any time of the year. So go ahead, buy Kato Grape Juice with Nata De Coco online today!. Grape Flavour",
    "disclaimer": "Despite our attempts to provide you with the most accurate information possible, the actual packaging, ingredients and colour of the product may sometimes vary. Please read the label, directions and warnings carefully before use.",
    "id": 100081004,
    "images": Array [
      "https://api.reliancesmart.in//images/?name=images/hbf/h68/9006373404702.jpg",
      "https://api.reliancesmart.in//images/?name=images/h2c/hbb/9006338408478.jpg",
      "https://api.reliancesmart.in//images/?name=images/h84/h32/9006353350686.jpg",
      "https://api.reliancesmart.in//images/?name=images/h5c/hae/9006365081630.jpg",
      "https://api.reliancesmart.in//images/?name=images/he2/h0b/9006370127902.jpg",
    ],
    "ingredients": undefined,
    "isFc": 1,
    "isVeg": null,
    "manufacturerAddress": "",
    "mrp": 60,
    "name": "Kato Grape Juice with Nata De Coco 330 ml",
    "nutritionalFacts": undefined,
    "skuId": 10092643,
    "sp": 58.8,
  },
  Object {
    "additionalInformation": "",
    "alias": "Valentino Sparkling Fruit Cocktail 750 ml 1 pcs",
    "bestBefore": 2020-11-02T06:30:00.000Z,
    "brandName": "Valentino",
    "category": Object {
      "id": 10001089,
      "name": "Fruit Juice",
    },
    "coupon": null,
    "deliverySlot": undefined,
    "description": "With zero percent alcohol, no colouring or preservatives, using only the highest quality ingredients, this beverage is ideal for parties, family celebrations, and picnics. Buy Valentino Sparkling Fruit Cocktail to enjoy the tingling sense of refreshment. It's enriched with the goodness of fruit, which is perfect for moments when you want to relax after a tiring day or wish for something refreshing.",
    "disclaimer": "Despite our attempts to provide you with the most accurate information possible, the actual packaging, ingredients and colour of the product may sometimes vary. Please read the label, directions and warnings carefully before use.",
    "id": 100068817,
    "images": Array [
      "https://api.reliancesmart.in//images/?name=images/h1c/h97/8932138745886.jpg",
      "https://api.reliancesmart.in//images/?name=images/h65/haf/8932134092830.jpg",
      "https://api.reliancesmart.in//images/?name=images/h32/hd6/8932138287134.jpg",
      "https://api.reliancesmart.in//images/?name=images/h67/h72/8932142022686.jpg",
      "https://api.reliancesmart.in//images/?name=images/h5b/hd7/9023370559518.jpg",
      "https://api.reliancesmart.in//images/?name=images/h5a/hac/9023397494814.jpg",
      "https://api.reliancesmart.in//images/?name=images/ha9/ha6/9023409487902.jpg",
    ],
    "ingredients": undefined,
    "isFc": 1,
    "isVeg": null,
    "manufacturerAddress": "",
    "mrp": 325,
    "name": "Valentino Sparkling Fruit Cocktail 750 ml",
    "nutritionalFacts": undefined,
    "skuId": 10078727,
    "sp": 318.5,
  },
]
`;
